### PR TITLE
Switch to using clangd with compile_flags.txt

### PR DESCRIPTION
**Pull request by ohodson**
A step towards retiring the problematic compile_commands.json.

WIP: part macOS, part Linux, no part Windows. Seems to work reasonably well on Linux / maxOS
TBD: per platform tweaks? e.g. choose CPU, arch, etc so more likely to have the appropriate code paths prettified.
TODO: update documentation, see if we can run generator steps rather than full build for flags.
TODO: external/

The kj convention of comments below rather than above limits some of the benefits here (e.g. typically shows documentation from preceding function rather than current function because kj puts comments below the declaration), but otherwise seems good.

Bug: https://github.com/cloudflare/workerd/issues/506

### CODE REVIEW

**Comment by ohodson** [comment]
Tried on Windows today, clangd that is packaged with VSCode (15.0.6). The symbolic links such as `bazel-bin` and `external` get resolved to their linked location and look for the standard clangd config files there and not in the workerd directory, e.g.
`external/` here links to `C:\tmp\rp4ezqc4\external` and the command:

```
C:\Users\Orion\src\workerd>  C:\Users\Orion\AppData\Roaming\Code\User\globalStorage\llvm-vs-code-extensions.vscode-clangd\install\15.0.6\clangd_15.0.6\bin\clangd.exe --check=external/v8/src/inspector/v8-regex.h
```
shows these config files being looked for:
```
clangd.exe	CreateFile	C:\.clangd	NAME NOT FOUND
clangd.exe	CreateFile	C:\tmp\.clangd	NAME NOT FOUND
clangd.exe	CreateFile	C:\tmp\rp4ezqc4\.clangd	NAME NOT FOUND
clangd.exe	CreateFile	C:\tmp\rp4ezqc4\external\.clangd	NAME NOT FOUND
clangd.exe	CreateFile	C:\tmp\rp4ezqc4\external\v8\.clangd	NAME NOT FOUND
clangd.exe	CreateFile	C:\tmp\rp4ezqc4\external\v8\src\.clangd	NAME NOT FOUND
clangd.exe	CreateFile	C:\tmp\rp4ezqc4\external\v8\src\inspector\.clangd	NAME NOT FOUND
clangd.exe	CreateFile	C:\Users\Orion\AppData\Local\clangd\config.yaml	PATH NOT FOUND
clangd.exe	CreateFile	C:\tmp\rp4ezqc4\external\v8\src\inspector\compile_commands.json	NAME NOT FOUND
clangd.exe	CreateFile	C:\tmp\rp4ezqc4\external\v8\src\inspector\build\compile_commands.json	PATH NOT FOUND
clangd.exe	CreateFile	C:\tmp\rp4ezqc4\external\v8\src\inspector\compile_flags.txt	NAME NOT FOUND
clangd.exe	CreateFile	C:\tmp\rp4ezqc4\external\v8\src\compile_commands.json	NAME NOT FOUND
clangd.exe	CreateFile	C:\tmp\rp4ezqc4\external\v8\src\build\compile_commands.json	PATH NOT FOUND
clangd.exe	CreateFile	C:\tmp\rp4ezqc4\external\v8\src\compile_flags.txt	NAME NOT FOUND
clangd.exe	CreateFile	C:\tmp\rp4ezqc4\external\v8\compile_commands.json	NAME NOT FOUND
clangd.exe	CreateFile	C:\tmp\rp4ezqc4\external\v8\build\compile_commands.json	PATH NOT FOUND
clangd.exe	CreateFile	C:\tmp\rp4ezqc4\external\v8\compile_flags.txt	NAME NOT FOUND
clangd.exe	CreateFile	C:\tmp\rp4ezqc4\external\compile_commands.json	NAME NOT FOUND
clangd.exe	CreateFile	C:\tmp\rp4ezqc4\external\build\compile_commands.json	PATH NOT FOUND
clangd.exe	CreateFile	C:\tmp\rp4ezqc4\external\compile_flags.txt	NAME NOT FOUND
clangd.exe	CreateFile	C:\tmp\rp4ezqc4\compile_commands.json	NAME NOT FOUND
clangd.exe	CreateFile	C:\tmp\rp4ezqc4\build\compile_commands.json	PATH NOT FOUND
clangd.exe	CreateFile	C:\tmp\rp4ezqc4\compile_flags.txt	NAME NOT FOUND
clangd.exe	CreateFile	C:\tmp\compile_commands.json	NAME NOT FOUND
clangd.exe	CreateFile	C:\tmp\build\compile_commands.json	PATH NOT FOUND
clangd.exe	CreateFile	C:\tmp\compile_flags.txt	NAME NOT FOUND
clangd.exe	CreateFile	C:\compile_commands.json	NAME NOT FOUND
clangd.exe	CreateFile	C:\build\compile_commands.json	PATH NOT FOUND
clangd.exe	CreateFile	C:\compile_flags.txt	NAME NOT FOUND
```

Prefixing the path with `./` doesn't work either. This thread looks relevant to the problem: https://github.com/clangd/clangd/issues/544

**Comment by kentonv** [comment]
Are the various header re-orderings actually related to this change? I can't tell what difference they would make.

**Comment by ohodson** [comment]
> Are the various header re-orderings actually related to this change? I can't tell what difference they would make.

Some, make clangd happier when looking at files standalone, others are just leftovers from trying and failing to make clangd happier in some way. This is only uploaded to make it easier to pass around Linux, macOS, Windows...

**Comment by fhanau** [comment]
LGTM overall, with the understanding that this is "complete enough" and will require changes in the future as flags change. One thing that might be difficult to handle well is the system root on Linux and Windows, which is currently dependent on a specific clang/MSVC version – not sure if there's an elegant way around that

File: tools/unix/create-external.sh [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+workspace=$(bazel info workspace)
+external="${workspace}/external"
+ln -sfF "${output_path}/../../../external" "${external}"
+
+# Temporary warning that compile_commands.json exists and will
+# interfere with the intended clangd setup.
+compile_commands="${workspace}/compile_commands.json"
+if [ -f "compile_commands.json" ] ; then
+  cat<<COMPILE_COMMANDS_WARNING
+WARNING: This workspace has a compile_commands.json file, but workerd
```
Comment (kentonv) [inline tools/unix/create-external.sh:17]: Let's add this check to the internal repo too! And delete all the hedron stuff.

**Review by jasnell — APPROVED** [review]

**Comment by fhanau** [comment]
There's still a few obsolete references to compile_commands: `.gitignore` checks for it and `docs/vscode.md` has both the new section where we explain why we stopped using it as well as an obsolete section indicating it's needed for clangd

**Comment by ohodson** [comment]
> There's still a few obsolete references to compile_commands: `.gitignore` checks for it and `docs/vscode.md` has both the new section where we explain why we stopped using it as well as an obsolete section indicating it's needed for clangd

Thanks, I've missed a chunk is docs/vscode.md and will fix. I think we should keep .gitignore as-is to avoid anyone checking compile_commands.json in.

**Comment by fhanau** [comment]
> > There's still a few obsolete references to compile_commands: `.gitignore` checks for it and `docs/vscode.md` has both the new section where we explain why we stopped using it as well as an obsolete section indicating it's needed for clangd
> 
> Thanks, I've missed a chunk is docs/vscode.md and will fix. I think we should keep .gitignore as-is to avoid anyone checking compile_commands.json in.

Thank you. I found one more reference at `.vscode/c_cpp_properties.json` for the mac build – not sure if that can be removed or it should be updated to refer to compile_flags.txt

**Comment by ohodson** [comment]
Thanks, I've uploaded https://github.com/cloudflare/workerd/pull/814 for `vscode.md`.

`.vscode/c_cpp_properties.json` only appears in the internal tree here. IIRC correctly that is for the Intellisense using Microsofts C/C++ extension: we should be able remove that file completely.

**Pull request merged**